lib/reservation.rb

```diff
@@ -0,0 +1,53 @@
+require 'date'
+
+
+require_relative 'frontdesk.rb'
+require_relative 'date.rb'
+
+module Hotel
+  class Reservation < DateRange
+    attr_accessor :start_date, :end_date, :num_rooms, :assigned_room, :discount, :block, :block_key
+    
+    COST = 200.00
+
+    def initialize(start_date:, end_date:, num_rooms:, discount: nil, block: :SINGLE, block_key: nil)
+      @start_date = start_date 
+      @end_date = end_date 
+      @num_rooms = num_rooms
+      @assigned_room = []
+      @discount = discount
+      @block = block
+      @block_key = block_key
+
+      raise ArgumentError.new("Invalid times: #{@start_date} comes after #{@end_date}") if (@start_date > @end_date || @start_date == @end_date)
+
+      raise ArgumentError.new("Invalid room request: provided #{@num_rooms} instead of an Integer") if @num_rooms.class != Integer
+
+      raise ArgumentError.new("Invalid room request: requested #{@num_rooms} rooms, #{@num_rooms-5} too many.") if @num_rooms > 5
+
+      raise ArgumentError.new("Invalid block key. Not valid for single room reservations") if @num_rooms == 1 && block_key != nil
+      
+      raise ArgumentError.new("Status and room argument: #{@num_rooms} room, should not be noted as a block") if (@num_rooms == 1 && @block == :BLOCK)
+
+      raise ArgumentError.new("Invalid status. Please choose: single or block.") unless [:SINGLE, :BLOCK].include? block
+    end
+
+    def total_cost
+      @num_rooms == 1 ? final_bill = (@end_date - @start_date)*@num_rooms*COST : final_bill = (@end_date - @start_date)*@num_rooms*COST*(1 - @discount)
+      return final_bill
+    end
```

Comment:
You seem very fond of ternary operators but you got a little carried away here:

```suggestion
    def total_cost
      if @num_rooms == 1
        return (@end_date - @start_date) * @num_rooms * COST
      else
        return (@end_date - @start_date) * @num_rooms * COST * (1 - @discount)
      end
    end
```

When code is complex breaking it up into multiple lines can help make it much easier to read.  

If you're going to do something like this as a ternary I'd recommend factoring out the assignment like this (since it can be easy to overlook and changes state):
```suggestion
    def total_cost
      final_bill = @num_rooms == 1 ? (@end_date - @start_date) * @num_rooms * COST : (@end_date - @start_date) * @num_rooms * COST * (1 - @discount)
      return final_bill
    end
```

In fact, looking at this code more makes it clear there's a lot of duplication between the two branches so it would be reasonable to use a ternary to calculate the multiplier though I'd still want to assign it to an intermediate variable:

```suggestion
    def total_cost
      multiplier = @num_rooms == 1 ? (1 - @discount) : 1
      return (@end_date - @start_date) * @num_rooms * COST * multiplier
    end
```